### DISCUSSION TITLE
Adding symbols from libge_*.a into blender's dynamic tables

### DISCUSSION THREAD

**Discussion by Liebranca** [UPBGE/upbge#1788] [Announcements & others]
Hi,

From the `ld` manual:

> If you use dlopen to load a dynamic object which needs to refer back to the symbols defined by the program, rather than some other dynamic object, then you will probably need to use this option when linking the program itself. 

The linker flag in question being `--export-dynamic`. With this, an extension could load symbols directly from blender, which I would appreciate.

Far as I can tell, this is  switched off. Tried:

```bash

nm --demangle --dynamic ~/blender-git/build_linux/bin/blender | grep "KX_GameObject::"

```

Which returns nothing. Running the same command on `libge_ketsji.a` I get a more informative answer: no dynamic symbols on any of the object files within the lib.

So: here I am, asking questions.

1. *where* do linker flags go?
2.  is `-fvisibility=hidden` switched on somewhere?

Reason why I want to do this:

```cpp

#include <iostream>
#include "KX_GameObject.h"

extern "C" void test(KX_GameObject* own) {
  KX_GameObject cpy=*own;
};

```

Linked as a shared object; loaded with ctypes at game startup, passed `cont.owner` as a `ctypes.py_object`, which AFAIK should correspond to `PyObject*`.

Running this test should attempt a vtable lookup, *then* segfault because I most likely passed it a bogus pointer. But instead it throws OSError as soon as I init my lib:

>bcpp-test.so: undefined symbol: _ZTV13KX_GameObject

And oh, that's too bad. But what does the binary say?

```
nm ~/blender-git/build_linux/bin/blender | grep "_ZTV13KX_GameObject"
0000000004ba5950 d _ZTV13KX_GameObject
```

Found. Let's demangle:

```
nm --demangle ~/blender-git/build_linux/bin/blender | grep "0000000004ba5950"
0000000004ba5950 d vtable for KX_GameObject
```

Looks to me me like Blender knows where the vtable is, it just doesn't want to share ;>

But as I see it, sharing is good. I don't even need every last symbol to be accessible to extensions, just the game engine ones for starters.

What do you guys think? Is this viable?

**Comment by youle31** [comment]
I don't understand anything sorry. It sounds very technical. I convert this to discussion as we are trying to keep this place for most important bugs only. Thanks for comprehension.

**Reply by Liebranca** [reply]
Oops! my bad.

Gist of it is I need to make a change in the build process to enable `dll` and `so` files to lookup functions within a running instance of blender or blenderplayer; that way compiled extensions wouldn't need to include them.

But this is for a C++ API so I don't think blenderdevs would tell me ;> Guess I'll have to try adding linker switches to game engine CMakeLists and see what happens.
